src/Storages/MergeTree/MergeTreePartition.cpp

```diff
@@ -14,11 +14,28 @@
 
 namespace DB
 {
+
 namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
 }
 
+namespace
+{
+    /// This is a special visitor which is used to get partition ID.
+    /// Calculate hash for UUID the same way as for UInt128.
+    /// It worked this way until 21.5, and we cannot change it,
+    /// or partition ID will be different in case UUID is used in partition key.
+    /// (It is not recommended to use UUID as partition key).
+    class LegacyFieldVisitorHash : public FieldVisitorHash
```

Comment:
I'm also thinking that copy-paste will be a better technique.
The LegacyFieldVisitorHash should be completely independent from FieldVisitorHash, so changing the latter will not affect it.